datasource/ethereum/src/ethereum_adapter.rs

```diff
@@ -978,20 +979,43 @@ where
         // Encode the call parameters according to the ABI
         let call_data = call.function.encode_input(&call.args).unwrap();
 
+        // Check if we have it cached, if not do the call and cache.
         Box::new(
-            // Make the actual function call
-            self.call(
-                logger,
-                call.address,
-                Bytes(call_data),
-                Some(call.block_ptr.number.into()),
-            )
-            .and_then(move |output| {
-                // Decode the return values according to the ABI
-                call.function
-                    .decode_output(&output.0)
-                    .map_err(EthereumContractCallError::from)
-            }),
+            match cache
+                .get_call(call.address, &call_data, call.block_ptr)
+                .map_err(|e| error!(logger, "call cache get error"; "error" => e.to_string()))
+                .ok()
+                .and_then(|x| x)
+            {
+                Some(result) => {
+                    Box::new(future::ok(result)) as Box<dyn Future<Item = _, Error = _> + Send>
+                }
+                None => {
+                    let cache = cache.clone();
+                    let call = call.clone();
+                    let call_data = call_data.clone();
+                    let logger = logger.clone();
+                    Box::new(
+                        self.call(
+                            &logger,
+                            call.address,
+                            Bytes(call_data.clone()),
+                            Some(call.block_ptr.number.into()),
+                        )
+                        .map(move |result| {
+                            let _ = cache
+                                .set_call(call.address, &call_data, call.block_ptr, &result.0)
+                                .map_err(|e| {
+                                    error!(logger, "call cache set error";
+                                                   "error" => e.to_string())
+                                });
+                            result.0
+                        }),
+                    )
+                }
+            }
+            // Decode the return values according to the ABI
+            .and_then(move |output| call.function.decode_output(&output).map_err(From::from)),
```

Comment:
I wonder if we should maybe only cache the result if wee succeed in decoding the outputs, to avoid caching invalid results caused e.g. by bugs in the Ethereum node. Because otherwise we might be forced to delete call results from the cache manually under certain situations.

Reply:
Sorry to bring this back but we got the problem with cached of invalid (failed) response from eth_call, and this behavior make retry meaningless and subgraph stuck at the block of eth_call failed. I think it is better to have the option to not save the result if the eth_call failed or cannot be decoded. Otherwise, there should be an easier way to clear the call_cached.

Reply:
hey @halink0803 - is this on the hosted service (if so can you share the specific subgraph and cached call), or is this on a local Graph Node?

Reply:
@azf20 it happened in both hosted service and my local graph-node also, because in the graph-node we call eth_call to rpc and sometimes, rpc returns with the bad result (empty or revert), here is the subgraph I got the issue on hosted service:  https://thegraph.com/hosted-service/subgraph/halink0803/kyberswap-exchange-bsc, I also tried on local graph-node and my solution was to clear the call_cache.

Reply:
@halink0803 call reverts appear in the logs, but for that subgraph I'm not seeing any call reverts logged.

Reply:
@leoyvens I don't get it. How the call reverts appear in the log but you are not seeing any call reverts logged? From the eth_call and the call_cache, it is actually show with empty response instead of throwing call error.

Reply:
How could I make the subgraph resume after this? Is this from subgraph code logic create the issue? It is showed eth_call reverted.

Reply:
Could you post the exact log line you are seeing, in the hosted service, indicating that a revert happened?